addons/survey/models/survey_survey.py

```diff
@@ -435,3 +489,40 @@ def action_survey_user_input_invite(self):
                     'search_default_invite': 1})
         action['context'] = ctx
         return action
+
+    @api.multi
+    def has_attempts_left(self, partner, email):
+        self.ensure_one()
+
+        if self.access_mode != 'public' and self.limit_number_of_attempts:
+            return self.attempts_left(partner, email) > 0
+
+        return True
+
+    @api.multi
+    def attempts_left(self, partner, email):
+        self.ensure_one()
+
+        domain = [
+            ('survey_id', '=', self.id),
+            ('test_entry', '=', False),
+            ('state', '=', 'done')
+        ]
+
+        if partner:
+            domain = expression.AND([domain, [('partner_id', '=', partner.id)]])
+        else:
+            domain = expression.AND([domain, [('email', '=', email)]])
+
+        return self.allowed_attempts_count - self.env['survey.user_input'].search_count(domain)
+
+    @api.multi
+    def _send_certification(self, user_input):
```

Comment:
Shouldn't this method be on `user_input`, as we can access `certificate` and `certification_email_template_id` fields through the `survey_id` field of `user_input`.

Reply:
You can also make the method callable in batch (for loop inside), as it returns nothing.

Reply:
Done.